server/manager/src/main/java/org/apache/accumulo/manager/Manager.java

```diff
@@ -960,39 +957,95 @@ private void checkForHeldServer(SortedMap<TServerInstance,TabletServerStatus> ts
+              removals.add(tableName);
+            }
+          });
+        }
+        removals.forEach(copy.getTableMap()::remove);
+        tserverStatusForLevel.put(k, copy);
       });
+      return tserverStatusForLevel;
+    }
+
+    private long balanceTablets() {
 
       final int tabletsNotHosted = notHosted();
       BalanceParamsImpl params = null;
       long wait = 0;
       long totalMigrationsOut = 0;
+      final Map<DataLevel,Set<KeyExtent>> partitionedMigrations =
+          partitionMigrations(migrationsSnapshot());
+
       for (DataLevel dl : DataLevel.values()) {
-        final Set<KeyExtent> migrationsForLevel = partitionedMigrations.get(dl);
-        if (migrationsForLevel == null) {
-          continue;
-        }
         if (dl == DataLevel.USER && tabletsNotHosted > 0) {
           log.debug("not balancing user tablets because there are {} unhosted tablets",
               tabletsNotHosted);
           continue;
         }
+        // Create a view of the tserver status such that it only contains the tables
+        // for this level in the tableMap.
+        final SortedMap<TServerInstance,TabletServerStatus> tserverStatusForLevel =
+            createTServerStatusView(dl, tserverStatus);
+        // Construct the Thrift variant of the map above for the BalancerParams
+        final SortedMap<TabletServerId,TServerStatus> tserverStatusForBalancerLevel =
+            new TreeMap<>();
+        tserverStatusForLevel.forEach((tsi, status) -> tserverStatusForBalancerLevel
+            .put(new TabletServerIdImpl(tsi), TServerStatusImpl.fromThrift(status)));
+
+        long migrationsOutCount = 0;
         long migrationsOutForLevel = 0;
-        int i = 0;
+        int attemptNum = 0;
         do {
-          i++;
-          log.debug("Balancing for tables at level {}, times-in-loop: {}", dl, i);
-          params = BalanceParamsImpl.fromThrift(tserverStatusForBalancer, tserverStatus,
-              migrationsForLevel);
+          log.debug("Balancing for tables at level {}, times-in-loop: {}", dl, ++attemptNum);
+          params = BalanceParamsImpl.fromThrift(tserverStatusForBalancerLevel,
+              tserverStatusForLevel, partitionedMigrations.get(dl));
           wait = Math.max(tabletBalancer.balance(params), wait);
-          migrationsOutForLevel = params.migrationsOut().size();
-          for (TabletMigration m : checkMigrationSanity(tserverStatusForBalancer.keySet(),
+          // The balancer may emit migrations for tables outside of the
+          // current level. We want to honor all migration requests, but
+          // need to keep track of the total for this level as well for
+          // the do-while loop condition.
+          migrationsOutCount = params.migrationsOut().size();
+          for (TabletMigration tm : params.migrationsOut()) {
+            if (dl == DataLevel.of(tm.getTablet().getTable())) {
+              migrationsOutForLevel++;
+            }
+          }
```

Comment:
May not need this change as we are now partitioning the input. This was part of an earlier commit.

Reply:
Removed in f0f0601